### DISCUSSION TITLE
Forward  search does not work with Sioyek.

### DISCUSSION THREAD

**Discussion by adwaraki** [James-Yu/LaTeX-Workshop#3730] [Q&A]
It appears that forward search does not work (or something seems to be broken) with Sioyek on macOS Ventura 13.2.1. Everything is at the latest build. I know this is not a supported configuration, but I am trying to figure it out since it is on the Wiki. I have exhausted my searching options (including whatever stuff ChatGPT and Bing chat have to throw at me (kudos, since they found some links even I could not find) and resorting to opening an issue as a last resort to see if someone on here can help me.

Points to note:

1. Disabling it and testing forward search with the built-in viewer works fine. The cursor indicates where it is supposed to be. No issues there.

2. Running the explicit command for forward-search on the command line with the same arguments in the json settings file works fine. Sioyek links to the exact line in the PDF and highlights it accordingly.  This leads me to believe there is something happening on the VSCode latex-workshop end. Here is the command line argument that does work:
```
sioyek --reuse-window --forward-search-file "/Users/XXXX/Research/Projects/Professional/Latex Projects/XXXX/XXXX.tex" --forward-search-line 33 "/Users/XXXX/Research/Projects/Professional/Latex Projects/XXXX/XXXX.pdf"
```

3. I initially thought the keybinding was not working, and changed it to something that was completely different, such as Ctrl+Cmd+Options+J (rather than the default Options+Cmd+J). Does not do anything.

4. I have also tried the alternate rebinding to Ctrl+Alt instead of the Cmd+Alt, and it still does not work.

5. I then thought this might have to do with selecting SyncTex from cursor in the left pane of latex-workshop, but I have done the same and it does not work.

6. Here are my external viewer settings on VSCode: 

```
"latex-workshop.view.pdf.external.viewer.command": "sioyek",
    "latex-workshop.view.pdf.external.synctex.command": "sioyek",
    "latex-workshop.view.pdf.external.synctex.args": [
        "--inverse-search",
        "code \"/Applications/Visual\\ Studio\\ Code.app/Contents/Resources\\app\\out\\cli.js\" --ms-enable-electron-run-as-node -r -g \"%1:%2\" ",
        "--reuse-window",
        "--forward-search-file",
        "%TEX%",
        "--forward-search-line",
        "%LINE%",
        "%PDF%"
    ],
```

Additional notes, in case someone comes across this from Google (or ChatGPT via Bing nowadays):

1. I struggled a lot to get inverse-search to work using the Wiki. A lot of tinkering later and the right settings are:

```
  "code \"/Applications/Visual\\ Studio\\ Code.app/Contents/Resources\\app\\out\\cli.js\" --ms-enable-electron-run-as-node -r -g \"%1:%2\" "
```
assuming you have the VSCode executable added to your path and `which code` actually returns the location of the executable, which typically should be `/Applications/Visual\ Studio\ Code.app/Contents/Resources/app/bin/code`. Please note the placement of the double quotes accordingly, since it is different on *nix, or could be different from what it is on the actual *nix settings on the Wiki.

2. The corresponding setting is `prefs.config` on the Sioyek side is `inverse_search_command 		code -r -g "%1:%2"`. That is it. It does not need the `cli.js` and other flags there.

3. Also, Sioyek does not have `--reuse-instance` as a flag, as mentioned on the Wiki. I am not sure if this is a difference between the Sioyek implementations on Windows and macOS, but it is `--reuse-window`. 

4. If it helps, I installed VSCode by downloading the package and Sioyek was installed using `brew install --cask sioyek` and the installation goes to `/opt/homebrew/cask/sioyek`.

**Comment by adwaraki** [comment]
The same issues exist with Skim, so that gives me to believe this is entirely related to keybindings and how latex-workshop is translating them internally. Inverse-search works fine with both Sioyek and Skim, so I guess Synctex is actually working properly. 

I am wondering if this is just a matter of how the keybindings are translated into the executable commands, or if it is something way simpler than that.

**Comment by James-Yu** [comment]
As external viewer is not a formally supported feature, I’m converting this issue to a discussion thread. Users may go on discussing over there.

**Comment by adwaraki** [comment]
It appears as if VSCode is actually receiving the correct keybindings. See keybinding debug traces below:

```
2023-02-18 01:57:24.943 [info] [KeybindingService]: / Received  keydown event - modifiers: [meta], code: MetaLeft, keyCode: 91, key: Meta
2023-02-18 01:57:24.945 [info] [KeybindingService]: | Converted keydown event - modifiers: [meta], code: MetaLeft, keyCode: 57 ('Meta')
2023-02-18 01:57:24.945 [info] [KeybindingService]: \ Keyboard event cannot be dispatched in keydown phase.
2023-02-18 01:57:24.945 [info] [KeybindingService]: / Received  keydown event - modifiers: [alt,meta], code: AltLeft, keyCode: 18, key: Alt
2023-02-18 01:57:24.945 [info] [KeybindingService]: | Converted keydown event - modifiers: [alt,meta], code: AltLeft, keyCode: 6 ('Alt')
2023-02-18 01:57:24.945 [info] [KeybindingService]: \ Keyboard event cannot be dispatched in keydown phase.
2023-02-18 01:57:25.441 [info] [KeybindingService]: / Received  keydown event - modifiers: [alt,meta], code: KeyJ, keyCode: 74, key: ∆
2023-02-18 01:57:25.441 [info] [KeybindingService]: | Converted keydown event - modifiers: [alt,meta], code: KeyJ, keyCode: 40 ('J')
2023-02-18 01:57:25.442 [info] [KeybindingService]: | Resolving alt+meta+[KeyJ]
2023-02-18 01:57:25.442 [info] [KeybindingService]: \ From 1 keybinding entries, matched latex-workshop.synctex, when: editorTextFocus && !config.latex-workshop.bind.altKeymap.enabled && !virtualWorkspace && editorLangId == 'latex', source: user extension James-Yu.latex-workshop.
2023-02-18 01:57:25.442 [info] [KeybindingService]: + Invoking command latex-workshop.synctex.
```
How do I figure out why it is not sending out the requisite command based on the command specified in the file?

**Comment by if-david** [comment]
Don't waste any more time—after trying at least 10 different solutions, the only one that works is installing the Full MacTeX
